buffer/src/main/java/io/netty5/buffer/internal/CleanerPool.java

```diff
@@ -0,0 +1,155 @@
+ *     Cleaner instance, so the pool won't be used in this case.</li>
+ * </ul>
+ */
+final class CleanerPool {
+    /**
+     * Our logger.
+     */
+    private static final InternalLogger logger = InternalLoggerFactory.getInstance(CleanerPool.class);
+
+    /**
+     * System property name used to configure the shared Cleaner pool size. External (non-event-loop) threads
+     * will always use the shared cleaner pool. The shared cleaner pool is instantiated lazily (only if it is
+     * needed).
+     * default=1.
+     * 0 means size will be set with the number of available processors.
+     */
+    private static final String CNF_POOL_SIZE = "io.netty5.cleanerpool.size";
+
+    /**
+     * System property name used to configure whether event-loop threads must use the shared cleaner pool.
+     * default=false, meaning that each event-loop thread will use its own cleaner instance
+     */
+    private static final String CNF_EVENT_LOOP_USE_POOL = "io.netty5.cleanerpool.eventloop.usepool";
+
+    /**
+     * Size of the shared Cleaner pool used by external threads, and optionally by event-loop threads.
+     * @see #CNF_POOL_SIZE
+     */
+    private static final int POOL_SIZE = getPoolSize(1);
+
+    /**
+     * Flag to configure whether event-loop threads must use the shared cleaner pool (default=false).
+     * @see #CNF_EVENT_LOOP_USE_POOL
+     */
+    private static final boolean EVENT_LOOP_USE_POOL = SystemPropertyUtil.getBoolean(CNF_EVENT_LOOP_USE_POOL, false);
+
+    /**
+     * Cleaner instances shared by all external threads, and optionally by event-loop threads
+     * These cleaners are wrapped in a static inner class for lazy initialization purpose.
+     */
+    private static class CleanersPool {
+        static final Cleaner[] cleaners = IntStream.range(0, POOL_SIZE)
+                .mapToObj(i -> createCleaner()).toArray(Cleaner[]::new);
+    }
+
+    /**
+     * A FastThreadLocal that returns a Cleaner to the caller thread.
+     * For event-loop threads, by default a dedicated cleaner instance is allocated and mapped to the
+     * caller thread, but by configuration, event-loop threads can use the shared cleaner pool if this is
+     * necessary.
+     * For external threads, a cleaner is always returned from the shared fixed cleaner pool in a round-robin way.
+     */
+    private final CleanerThreadLocal threadLocalCleaner = new CleanerThreadLocal();
+
+    /**
+     * The singleton for the CleanerPool.
+     */
+    public static final CleanerPool INSTANCE = new CleanerPool();
+
+    /**
+     * The FastThreadLocal used to map a Cleaner instance to Event Loop threads.
+     */
+    private static class CleanerThreadLocal extends FastThreadLocal<Cleaner> {
+        private static final AtomicInteger counter = new AtomicInteger();
+
+        @Override
+        protected Cleaner initialValue() {
+            if (!EVENT_LOOP_USE_POOL && ThreadExecutorMap.currentExecutor() != null) {
```

Comment:
How this is done it will also be true for `EventExecutor` implementations not only `EventLoop`s, which is not a bad thing. I just wonder if we should make this clear somehow. Also another thing we could do is just check for `FastThreadLocalThread`

Reply:
the intent was to check event loop threads and I was thinking that `ThreadExecutorMap.currentExecutor() != null` was checking only for event loop threads.

I have corrected by `Thread.currentThread() instanceof FastThreadLocalThread`, please double check ?